[DRAFT] feat: API

66Bunz opened the pull request:
# API Implementation

<details><summary>Roadmap</summary>
  <p>

The roadmap I have in mind for this project is this:
- **develop API based on current app implementations (this PR)**
- move all implementations from the app to the API making them more "agnostic" (not so view driven)
- change app to use API backend manipulating its data
- since API is available, more "plugins", "integrations", "providers" can be implemented between various services (Jellyfin, seer, Trakt, ecc ecc)

  </p>
</details>

---
Fixes #488
Fixes #543
Fixes #582
Fixes #698

## To-Dos

- [x] Basic Media Endpoints
  - [x] `/api/v1/calendar/`
    - [x] GET: List upcoming calendar events
  - [x] `/api/v1/calendar/update/`
    - [x] POST: Update calendar events
  - [x] `/api/v1/changes_history/{media_type}/{history_id}`:
    - [x] DELETE: Delete specific changes history entry
    - [x] GET: Get specific changes history entry
  - [x] `/api/v1/lists/`
    - [x] GET: List all user & shared lists
    - [x] POST: Create new list
  - [x] `/api/v1/lists/{list_id}/`
    - [x] DELETE: Delete specific list
    - [x] GET: Get specific list
    - [x] PATCH: Edit list properties (maybe also content)
  - [x] `/api/v1/lists/{list_id}/items/`
    - [x] GET: List items in list
  - [x] `/api/v1/lists/{list_id}/items/{item_id}/`
    - [x] DELETE: Delete item from list
    - [x] GET: Get tracked item
  - [x] `/api/v1/media/`
    - [x] GET: List tracked media
  - [x] `/api/v1/media/{media_type}/`
    - [x] GET: List tracked media of type
    - [x] POST: Track media of type
  - [x] `/api/v1/media/{media_type}/{source}/{media_id}/`
    - [x] DELETE: Delete tracked item
    - [x] GET: Get tracked item
    - [x] PATCH: Edit last consumption of tracked item
  - [x] `/api/v1/media/{media_type}/{source}/{media_id}/changes_history/`
    - [x] GET: Get changes history of tracked item
  - [x] `/api/v1/media/{media_type}/{source}/{media_id}/history/`
    - [x] GET: Get consumption history of tracked item
  - [x] `/api/v1/media/{media_type}/{source}/{media_id}/history/{consumption_id}/`
    - [x] DELETE: Delete specific consumption entry
    - [x] GET: Get specific consumption entry
    - [x] PATCH: Edit specific consumption entry
  - [x] `/api/v1/media/{media_type}/{source}/{media_id}/lists/`
    - [x] GET: Lists where the tracked item is in
  - [x] `/api/v1/media/{media_type}/{source}/{media_id}/lists/{list_id}`
    - [x] DELETE: Delete tracked item from specific list
    - [x] PUT: Add tracked item to specific list
  - [x] `/api/v1/media/{media_type}/{source}/{media_id}/recommendations/`
    - [x] GET: Get recommendations based on tracked item
  - [x] `/api/v1/media/{media_type}/{source}/{media_id}/seasons/`
    - [x] GET: List seasons of tracked tv serie if media is tv serie
  - [x] `/api/v1/media/{media_type}/{source}/{media_id}/sync/`
    - [x] POST: Update metadata of tracked item
  - [x] `/api/v1/media/{media_type}/{source}/{media_id}/{season}/`
    - [x] DELETE: Delete tracked season if media is tv serie
    - [x] GET: Get tracked season if media is tv serie
    - [x] PATCH: Edit last consumption of tracked season
  - [x] `/api/v1/media/{media_type}/{source}/{media_id}/{season}/episodes/`
    - [x] GET: List episodes of tracked season if media is tv serie
  - [x] `/api/v1/media/{media_type}/{source}/{media_id}/{season}/changes_history/`
    - [x] GET: Get changes history of tracked season
  - [x] `/api/v1/media/{media_type}/{source}/{media_id}/{season}/history/`
    - [x] GET: Get consumption history of tracked season
  - [x] `/api/v1/media/{media_type}/{source}/{media_id}/{season}/history/{consumption_id}/`
    - [x] DELETE: Delete specific consumption entry
    - [x] GET: Get specific consumption entry
    - [x] PATCH: Edit specific consumption entry
  - [x] `/api/v1/media/{media_type}/{source}/{media_id}/{season}/lists/`
    - [x] GET: Lists where the tracked item is in
  - [x] `/api/v1/media/{media_type}/{source}/{media_id}/{season}/lists/{list_id}`
    - [x] DELETE: Delete tracked item from specific list
    - [x] PUT: Add tracked item to specific list
  - [x] `/api/v1/media/{media_type}/{source}/{media_id}/{season}/sync/`
    - [x] POST: Update metadata of tracked season
  - [x] `/api/v1/media/{media_type}/{source}/{media_id}/{season}/{episode}/`
    - [x] DELETE: Delete tracked episode if media is tv serie
    - [x] GET: Get tracked episode if media is tv serie
    - [x] PATCH: Edit last consumption of tracked episode
  - [x] `/api/v1/media/{media_type}/{source}/{media_id}/{season}/{episode}/changes_history/`
    - [x] GET: Get changes history of tracked episode
  - [x] `/api/v1/media/{media_type}/{source}/{media_id}/{season}/{episode}/history/`
    - [x] GET: Get consumption history of tracked episode
  - [x] `/api/v1/media/{media_type}/{source}/{media_id}/{season}/{episode}/history/{consumption_id}/`
    - [x] DELETE: Delete specific consumption entry
    - [x] GET: Get specific consumption entry
    - [x] PATCH: Edit specific consumption entry
  - [x] `/api/v1/media/{media_type}/{source}/{media_id}/{season}/{episode}/lists/`
    - [x] GET: Lists where the tracked item is in
  - [x] `/api/v1/media/{media_type}/{source}/{media_id}/{season}/{episode}/lists/{list_id}`
    - [x] DELETE: Delete tracked item from specific list
    - [x] PUT: Add tracked item to specific list
  - [x] `/api/v1/media/{media_type}/{source}/{media_id}/{season}/{episode}/sync/`
    - [x] POST: Update metadata of tracked episode (season)
  - [x] `/api/v1/search/{media_type}/`
    - [x] GET: Search for media using the specified provider
  - [x] `/api/v1/statistics/`
    - [x] GET: Get user statistics
- [x] Yamtrack Endpoints
  - [x] `/api/v1/health/`
    - [x] GET: Health check endpoint
  - [x] `/api/v1/info/`
    - [x] GET: Get application info and version
- [x] Merge latest changes from `feat/add-api` branch
- [ ] Merge latest changes from `dev` branch
- [x] Error handling
- [x] Tests
- [ ] Documentation

## Disclaimer

This is my first time working with Django, so I'm not really familiar with all the best practices, I tried to follow the existing code style.

If someone has suggestions on how to improve something let me know!

## Main changes

### Implementations

Since all of the features of the api are already implemented for the WebApp, I just referenced basically all the existing methods from the `app` [views](https://github.com/66Bunz/Yamtrack/tree/feat/add-api/src/app/views.py) copying the implementations, or referencing directly the methods used in the `app` implementations.

Basically nothing was duplicated, only the [changes_history_processor](https://github.com/66Bunz/Yamtrack/tree/feat/add-api/src/api/changes_history_processor.py), because there were some things that were implemented to be human-readable, but not as api-friendly.

In the future, if there will be the decision that the WebApp should use the API directly, all the implementations should be moved from the `app` to the `api`.

### Urls

For urls I used the prefix `/api/v1/` just to be on the safe side in case of future updates of the api.

I choose to use the re_path in the [urls.py](https://github.com/66Bunz/Yamtrack/tree/feat/add-api/src/api/urls.py) to allow the ending with or without a trailing slash (both `/api/v1/media/` and `/api/v1/media` are accepted). This makes the paths a bit more complex to write, but I think it's more friendly for the end user.

### /media endpoint

The /media endpoint by default excludes seasons and episodes from the list of tracked items, to declutter the results.

### Specific media endpoints

The main change I did in structuring the API routes for media items is inverting the media type and the source, because I think it makes more sense to group by media type first. So instead of `/{source}/{media_type}/{media_id}/` as in the WebApp, I used `/{media_type}/{source}/{media_id}/`.

It's true that both the media type can have different sources and a source can have differrent media types, but mainly the media type is the primary categorization of tracked items (e.g., when listing tracked items, filtering, searching, etc.), so it makes more sense to have it first in the path.

I decided that for the media types to follow a parent-child relationship, so seasons and episodes are sub-resources of tv series, so `media_type` in the request URL cannot be `season` or `episode`, but only `tv`, `movie` ecc ecc. This was made to simplify the API structure and make it more intuitive to use. So to query seasons or episodes, the user has to specify the parent tv serie first (e.g., `/api/v1/media/tv/{source}/{media_id}/{season}/{episode}/`).

#### Handling multiple consumptions

A single consumption entry is identified by `/history/{consumption_id}/` subpath, so to get/edit/delete a specific consumption entry, the user has to specify the consumption id.

The consumption id is the id of the media in the database, so it's incremental between all medias of the same type, and it's not related to the media id of the tracked item, because a tracked item can have multiple consumption entries, so they need to be identified by a unique id.

When listing the consumption history of an item, all the consumption entries are returned with their consumption id, so the user can use that id to get/edit/delete a specific consumption entry.

#### Operations on specific media items

GET on /media/{media_type}/{source}/{media_id}/ returns the tracked item with all its data.
DELETE on /media/{media_type}/{source}/{media_id}/ deletes the tracked item completely, with all its consumptions.
PATCH on /media/{media_type}/{source}/{media_id}/ edits the last consumption of the tracked item.

All of these operations are available for single consumptions, to get/delete/edit a specific consumption entry, the user has to use the /history/{consumption_id}/ subpath.

All of the previous operations are available for seasons and episodes as well.

#### Changes history endpoints

A thing to note is that history entries of children can appear in the changes history of the parent item when querying the parent changes history endpoint. For example, when getting a changes history entry of a season by id, that same id is present when querying the changes history entry of the tv serie, but this means that the item id of the entry is not the correct one because it was built for the parent and not the child.

### Lists

Since the current implementation of lists doesn't have a field to identify the element in the list (0 for the first element added to the list, 1 for the second one and so on), I added a new field `list_item_id` to the CustomListItem model, which is the id of the item in the list, so it can be used to identify the item in the list and perform operations on it (e.g., delete a specific item from the list).

I implemented also a logic to update the `list_item_id` of the items when items are removed, to not have "gaps" in the numeration. This change could be leveraged in the future to implement custom sorting for lists (out of scope for this PR).

### Sorting

Sorting for media items was implemented only partially, because there already are some ways to sort items in the existing codebase functions, so I just referenced those functions.

For lists, I had to implement them from scratch, because there were no existing implementations. They are implemented in the [helpers.py](https://github.com/66Bunz/Yamtrack/tree/feat/add-api/src/api/helpers.py) file.

Probably in the future it could be useful to re-implement sorting for media items too, to have more control over it. At that point probably a new sorting.py file could be created to handle all the sorting logic in one place.

### Search

The search is only implemented for providers by type, so the user has to specify the media type in the url, and cannot search only between already tracked items.

Search should be done globally, returning tracked and not tracked elements, maybe adding a tracked boolean field in the element, it'll come in the future.

Also Search should support more filters in the future, and also not be binded to media type, but a global search with elements from all providers could be implemented.

<!-- For manual items, search can only be implemented between all tracked items, as there is no external source to search from. -->

### Statistics

Right now the statistics endpoint returns exactly the same data as the WebApp statistics page, without any modification. This is not really the best thing as they should be computed specifically for the API, so they can be computed as the API implementer wants. I wasn't really sure what to include, so I just left it as is for now.

### Episode sync

Since in the `app` the episode sync method was not implemented, I just referenced the season sync method for now. In the future, if there will be the need to have a specific episode sync implementation, it can be changed easily.

### Authentication

The API uses two authentication methods implemented in [authentication.py](https://github.com/66Bunz/Yamtrack/tree/feat/add-api/src/api/authentication.py):

- **BearerAuthentication**: `Authorization: Bearer <token>` - Standard OAuth-style bearer tokens
- **APIKeyAuthentication**: `X-API-Key: <token>` - Custom header for simpler API key auth

Both methods look up the user by the `token` field in the User model. The user's existing token (from the users app) is reused for API authentication. No separate API key generation was implemented yet.

**Note**: Currently there's no way to generate or regenerate API tokens via the API itself - users must use the web interface.

### Pagination

All list endpoints use a consistent pagination approach via `parse_limit_offset` and `paginate_data` helpers in [helpers.py](https://github.com/66Bunz/Yamtrack/tree/feat/add-api/src/api/helpers.py):

- Query params: `limit` (default 20, max 200) and `offset` (default 0)
- Response includes `pagination` object with `total`, `limit`, `offset`, `next`, and `previous` URLs
- The `next` and `previous` URLs preserve all original query params (filters, sort, search)

**Consideration**: This is offset-based pagination which can have performance issues with big lists of items.

### Error Handling

A custom middleware [ApiJsonErrorMiddleware](https://github.com/66Bunz/Yamtrack/tree/feat/add-api/src/api/middleware.py) converts Django HTML error responses to JSON for all `/api/` paths:

- Intercepts 4xx/5xx responses and ensures they return JSON with `{"detail": "message"}` structure
- Also has `process_exception` to catch unhandled exceptions and return JSON errors
- In DEBUG mode, includes additional debug information in the response

This is because Django by default returns HTML error pages, but APIs shouldn't return HTML data.

### Serializers

I added custom serializers in [serializers.py](https://github.com/66Bunz/Yamtrack/tree/feat/add-api/src/api/serializers.py) to format the API responses consistently, each resource (Media, History, Season, Episode, List, etc.) has its own serializer class,  nested relationships (e.g., Media with Seasons and Episodes) are handled correctly.

### Testing

Still missing

### Known Issues & Considerations

1. **Statistics endpoint**: Returns exactly what the web statistics page needs, not necessarily what an API consumer would want. Should be redesigned with API-first thinking.
2. **No API token management**: Users can't generate/regenerate/revoke API tokens via API - must use web interface.
3. **Search limitations**: Can't search seasons/episodes (intentionally). Can't search already tracked media (only provider search).
4. **No rate limiting**: Currently no rate limiting implemented, could lead to abuse.
5. For children items, errors should shown for non existing season/episode (for example if it's requested season 4 of a 2 season show)
6. Should the list of all media contain all the episodes and seasons? They can be hidden by default and only activated by a flag?

### Future Improvements

1. **Field filtering**: Allow clients to request only specific fields (`?fields=title,score`)
2. **Batch operations**: Add support for bulk operations (e.g., update multiple items in one request)
3. **Webhooks**: Event subscription for calendar updates, new releases
4. **Rate limiting**: Proper per-user rate limiting with headers (X-RateLimit-*)
5. **Expand filters**: More granular filtering (score range, date range, tags)
6. **Sorting**: More sort options and multi-field sorting directly on the db query
7. **Provider integrations**: Endpoints to import tracked media from external providers ([#1069](https://github.com/FuzzyGrim/Yamtrack/pull/1069#issuecomment-3707953738))
8. **Improve media user fields**: Allow setting notes, progress, status, score and dates to every media, because not every media supports it, see table below.
9. **Custom lists ordering**: Allow users to create a custom order to the lists.
10. **Administration commands**: Add administration commands

    <details><summary>table</summary>
      <p>

      | Media Type | score | status | progress | start_date | end_date | notes |
      |---|---|---|---|---|---|---|
      | Movie | ✅ | ✅ | ❌ | ✅ | ✅ | ✅ |
      | TV | ✅ | ✅ | ❌ | ❌ | ❌ | ✅ |
      | Season | ✅ | ✅ | ❌ | ❌ | ❌ | ✅ |
      | Episode | ❌ | ❌ | ❌ | ❌ | ✅ | ❌ |
      | Anime | ✅ | ✅ | ✅ | ✅ | ✅ | ✅ |
      | Manga | ✅ | ✅ | ✅ | ✅ | ✅ | ✅ |
      | Game | ✅ | ✅ | ✅ | ✅ | ✅ | ✅ |
      | Book | ✅ | ✅ | ✅ | ✅ | ✅ | ✅ |
      | Comic | ✅ | ✅ | ✅ | ✅ | ✅ | ✅ |

      </p>
    </details>
11. Implement admin commands to manage the server

### Dependencies

Only `djangorestframework` was added as a dependency, no additional dependencies were needed.

66Bunz committed:
Added prelimiary draft of a possible API spec

66Bunz commented:
@FuzzyGrim is it possible to create a branch `feat/add-api`?

FuzzyGrim commented:
Hi @66Bunz,

Thanks for working on this!

The spec file looks good to me. I've created the `feat/add-api` branch.

Also, I'd prefer a separate `api` app. Do you plan to use the Django REST framework for the implementation?

66Bunz commented:
Thanks!
Yes, that's what I've been using

66Bunz committed:
Merge branch 'FuzzyGrim:feat/add-api' into feat/add-api

66Bunz committed:
feat(api): Started implementing media management API with authentication and pagination

- Added Bearer and API Key authentication classes for secure access
- Created serializers for media and item details
- Started implementing API endpoints copying `app` implementations (could be useful to change some implementations in the future)

66Bunz committed:
feat(api): Added calendar endpoints

66Bunz committed:
Add middleware to return json errors instead of html ones.

66Bunz committed:
feat(api): Refactor URL patterns to use re_path for improved handling

66Bunz commented:
I've got an open questions:

Looking at [this list](https://github.com/FuzzyGrim/Yamtrack/pull/924#issue-3538082363), is it ok to have the `/api/v1/media/[media_type]/[source]/[media_id]/history/[history_id]` route to delete an history record as it's now, or it's better to have it at `/api/v1/history/[history_id]`?

The history_id is not bound to the specific media

66Bunz committed:
feat(api): Refactor media handling and history processing

- Introduced new `EpisodeSerializer` for handling episode-specific data
- Updated `MediaSerializer` to include `parent_id` for nested relationships (tv serie -> season -> episode)
- Updated `helpers.py` to start to support episodes
- Added `history_processor.py` for managing history entries and processing timeline data
- Continued implementation of routes
- Modified URL patterns to accommodate new endpoints for media recommendations and history details
- Improved pagination and serialization logic for media and history responses

FuzzyGrim commented:
Another option would be `/api/v1/history/{media_type}/{history_id}`. Since `history_id` is unique within its media type.

66Bunz commented:
@FuzzyGrim I've got a question: Is it possible to get the media_id that is related to the history entry?

I looked around but I wasn't able to find anything useful...

66Bunz committed:
feat(api): Implement history management endpoints and serializers

66Bunz committed:
feat(api): Add seasons retrieval endpoint and update schemas for media seasons

66Bunz committed:
feat(api): Reorder endpoint classes

66Bunz committed:
feat(api): Implement media sync endpoint

66Bunz committed:
feat(api): Add media season detail, episodes, and history endpoints; update pagination and schemas

66Bunz committed:
feat(api): Add sync endpoint for media season metadata

FuzzyGrim commented:
Something like this:

```
historical_model = apps.get_model("app", f"historical{media_type}")
record = historical_model.objects.get(history_id=history_id)

media_model = apps.get_model("app", media_type)
media_instance = media_model.objects.get(id=record.id)
media_id = media_instance.item.media_id
```

66Bunz commented:
> Something like this:
> 
> ```
> historical_model = apps.get_model("app", f"historical{media_type}")
> record = historical_model.objects.get(history_id=history_id)
> 
> media_model = apps.get_model("app", media_type)
> media_instance = media_model.objects.get(id=record.id)
> media_id = media_instance.item.media_id
> ```

@FuzzyGrim 1 thing I noticed is that if the history entry is of a specific season/episode, I can only get the serie out of it, I wasn't able to get the specific season/episode. Am I missing something or this is it?

FuzzyGrim commented:
It should work for both seasons and episodes. The `media_type` wouldbe `episode` or `season`. But tv shows, seasons and episodes all share the same `media_id`. TMDB doesn't have a separate unique identifier for each, but the `media_instance.item` will include `season_number` and `episode_number` in it.

66Bunz commented:
Yes, nevermind, I was doing it wrong

66Bunz committed:
feat(api): Add endpoints for media episode details, history, and sync; update serializers and helpers

66Bunz commented:
@FuzzyGrim Right now I'm only missing the creation and deletion of medias and everything about lists, then I'll remove the draft status from the PR.
The only thing you need to do then is to make the feat/api branch up to date with dev, so I can add the latest changes made in dev to the api, and then this can be merged.
If you have comments, directives I'll be happy to hear them.

66Bunz commented:
I'll write some documentation for the api here, to explain what I did and why I did it, so you can understand better the thought process behind this, and it could be useful to review the changes.
If you want/think this could add value to the project the next thing I want to do after finishing this feature is probably make a better docs system using GH pages, so the full API docs could be written, if you have some directives on that I'll create an issue for it in the next future.

66Bunz committed:
feat(api): Fix response status for media sync endpoints to 429 instead of 409

66Bunz committed:
feat(api): Fix episode history summary

66Bunz commented:
@FuzzyGrim I noticed in the history entry a `change_reason` field, but it looks like that field is always empty. Is it used for something I haven't encountered?

66Bunz committed:
feat(api): Update docstrings of views and functions

66Bunz committed:
feat(api): Start refactor of HTTP response messages using centralized helper

66Bunz committed:
feat(api): Improve docstrings

66Bunz committed:
feat(api): Refactor middleware for improved error handling and response conversion

66Bunz committed:
feat(api): Enhance helpers

FuzzyGrim commented:
Synced `feat/api` with `dev`.

I've never tried it, but I believe it's possible to automate a GitHub page to keep the documentation up-to-date using GitHub Actions.

Currently `change_reason` is not used, it was automatically generated by [django-simple-history](https://django-simple-history.readthedocs.io/en/stable/)

66Bunz commented:
> Synced `feat/api` with `dev`.

Thanks

> I've never tried it, but I believe it's possible to automate a GitHub page to keep the documentation up-to-date using GitHub Actions.

It's possible to serve docs directly from github by github, only a `docs` folder with some markdown files and something more is needed in the project root, and from the repo settings you publish the folder as [github pages](https://docs.github.com/en/pages). Only if you want to use an external documentation provider you'd use github actions.

66Bunz committed:
feat(api): Refactor API views for improved serialization and reuse of helper functions

- Updated media fetching logic in MediaListView and MediaTypeListView to use new helper functions.
- Simplified response serialization across various views using a unified serialize_data function.
- Improved handling of media history and seasons in MediaHistoryView and MediaSeasonsView.
- Removed redundant code and comments for clarity.

66Bunz committed:
feat(api): Rename UpdateCalendarView to CalendarUpdateView for consistency

66Bunz commented:
Added some infos on the changes in the [top comment](https://github.com/FuzzyGrim/Yamtrack/pull/924#issue-3538082363)

66Bunz committed:
feat(api): Implement delete functionality for media, seasons, and episodes

66Bunz committed:
feat(api): Update API schema to support multiple watches

66Bunz committed:
feat(api): Initial implementation to manage multiple watches of a single media.

- Renamed old `/history` endpoints to `/changes_history` and updated function definitions/docs
- Media details now show the multiple watches of the media with their data

66Bunz committed:
feat(api): Fix classes names

66Bunz committed:
feat(api): Reorder classes

66Bunz committed:
feat(api): Add health check endpoint and response serializer

66Bunz committed:
feat(api): Add info endpoint and response serializer

66Bunz committed:
feat(api): Refactor media consumption history handling

- Introduced `get_progress_from_status` helper function to determine progress based on media status.
- Added `ChangesHistoryEntrySerializer` to handle serialization of change-based history entries.
- Refactored `CompleteEpisodeSerializer` and `CompleteMediaSerializer` to use consumptions instead of watches.
- Implemented new views for media consumption history and entry detail, including deletion and retrieval of specific consumption entries.
- Updated URL patterns to include routes for media consumption history and entry detail.
- Adjusted existing views to utilize the new `ChangesHistoryEntrySerializer` for consistency.

66Bunz committed:
feat(api): Added support for updating user data of already tracked media

66Bunz committed:
feat(api): Forgotten changes from the previous commit

66Bunz committed:
chore(migrations): remove obsolete migration

66Bunz committed:
feat(api): Update identification of consumptions.

Now they aren't identified with increasing ids added at runtime, but the global consumption ids of the Yamtrack db are used.

66Bunz commented:
@FuzzyGrim is the fact that one can add an untracked media to a list a bug or intended?

Because if it's intended I'll have to change the entire logic of the lists handling

FuzzyGrim commented:
It's intended

66Bunz committed:
feat(api): Update API spec

66Bunz committed:
feat(api): Add lists endpoint

66Bunz committed:
feat(api): Enhanced lists endpoints

- Fix serializers to support untracked items serialization
- Added a couple of missing fields in serialized data
- Added list_item_id field to CustomListItem
- Updated lists tests

66Bunz committed:
feat(api): Add missing consumption_id field to Media schema

66Bunz committed:
feat(api): Implement list membership retrieval for media items

66Bunz committed:
feat(api): Update API to include lists retrieval for medias

66Bunz committed:
feat(api): Add pagination to media lists endpoints and add endpoints for managing medias in lists

66Bunz commented:
The end is near!! 🥳😅

Only a couple of endpoints needs to be written, then I'm off to solve a couple of TODOs I set myself, and then I can start merging the latest changes from the 21st of October from the `dev` branch.

After that the draft status can be removed and the PR can be merged!!

66Bunz committed:
feat(api): Remove unused permissions module

66Bunz committed:
feat(api): Remove some ruff errors

66Bunz committed:
feat(api): Refactor pagination logic to use helper function in SearchProviderView

66Bunz committed:
feat(api): Refactor error responses for better readability

66Bunz committed:
feat(api): Remove TODO of previous commit

66Bunz committed:
feat(api): Integrate search functionality in lists.models.get_user_lists

austin-ek commented on src/api/helpers.py:
```diff
@@ -0,0 +1,682 @@
+    if not status:
+        return MediaStatusChoices.ALL
+    try:
+        return get_media_status(int(status))
```
status filtering is broken, this fixes it


```suggestion
        return get_media_status(int(status), reverse=True)
```

66Bunz replied:
I know some sorting it's broken, but I'm not sure this will fix it everywhere.
I wanted to rewrite it and move it to the models to improve the performances tho, so I'll probably rewrite it from scratch, I just have to fix the Episode serializer that is not working right now.
Thanks for the help tho!! <3

austin-ek replied:
thanks! and thanks for all of your work, i'm looking forward to this being merged!

66Bunz committed:
feat(api): Rename get_item_lists_payload to get_item_lists for clarity

66Bunz committed:
feat(api): Uniformed endpoint responses for lists of seasons in a serie and episodes in a season to the Media object

66Bunz committed:
feat(api): Update some errors in the spec from https://github.com/FuzzyGrim/Yamtrack/issues/488#issuecomment-4089002744

66Bunz committed:
feat(api): Implement search functionality for manually created media with pagination support

66Bunz committed:
feat(api): Fix new media tracking

fixed an error in providers/manual.py that returned related.seasons object even if the media type wasn't tv.

66Bunz commented:
@FuzzyGrim is it possible to make `feat/add-api` up to date with `dev`?
I'm ready to merge the changes made in the last months to my branch to finally start removing some conflicts

66Bunz committed:
Merge remote-tracking branch 'upstream/feat/add-api' into feat/add-api

66Bunz committed:
feat(api): Update API spec and some Ruff errors

66Bunz committed:
feat(tests): Add first API tests for authentication, calendar, and changes history

66Bunz committed:
Merge upstream/feat/add-api into feat/add-api

66Bunz committed:
feat(api): Add BoardGame API support

66Bunz committed:
feat(tests): Add tests for public API endpoints

66Bunz committed:
feat(api): Enhance date range resolution in CalendarView

66Bunz committed:
feat(tests): Refactor API test cases to use YamtrackApiTestCase and enhance them

66Bunz committed:
feat(api): Add tests for changes history

66Bunz committed:
feat(tests): Add statistics tests

66Bunz committed:
feat(api): Fix media deletion from list

66Bunz committed:
feat(api): Change ItemSerializer to return numeric media_id instead of string.

66Bunz committed:
feat(tests): Add some media core tests for API endpoints

66Bunz committed:
feat(api): Add 404 response for media not found in MediaConsumptionHistoryView

66Bunz committed:
feat(api): Add missing response schema for sync endpoint in OpenAPI specification

66Bunz committed:
feat(tests): Add more media core API tests

66Bunz committed:
feat(tests): Update mock metadata for TV media items with detailed structure and recommendations

66Bunz committed:
feat(tests): Add tests for seasons

66Bunz committed:
feat(api): Improve date validation in _validate_dates function

66Bunz committed:
feat(tests): Add tests for media episodes

66Bunz committed:
feat(tests): Add tests for lists

66Bunz committed:
feat(tests): Add tests for api search

66Bunz committed:
feat(tests): Add complete lists structure validation in test helpers

66Bunz committed:
Merge remote-tracking branch 'upstream/feat/add-api' into feat/add-api

66Bunz committed:
feat(tests): Update ruff rules and fixed some errors (200->72)

66Bunz committed:
fix(helpers): Simplify first_day assignment using logical OR

66Bunz committed:
Merge remote-tracking branch 'upstream/feat/add-api' into feat/add-api

66Bunz committed:
feat(api): Add missing sorting functionality to apply_manual_sort_for_type

66Bunz committed:
feat(api): Improve sorting functionality for items in apply_list_sort

66Bunz committed:
feat(api): Update media_id handling to support alphanumeric values across serializers, views, and tests

66Bunz committed:
feat(tests): Update media_id assertion to validate new UUID format in media core tests

66Bunz committed:
feat(api): Enhance CustomList serialization with item count and latest update fields

66Bunz committed:
feat(api): Normalize search query and improve caching for manual media search

66Bunz committed:
feat(api): Initialize episode count in metadata for TV and season media types

mbuet2ner commented:
Really love the API! Makes a lot of things possible, e. g. mobile apps. Created a (yes, fully vibe-coded) iOS app [here](https://github.com/mbuet2ner/yamtrack-ios). Swift 6, iOS 26, Liquid Glass, decent test coverage.

Wanted to have an easier way to add media "on-the-go" and tried I far I can get with codex. Looking forward to the official release of the API!

Haggler7039 commented:
I've had some more time to play around with the API, I found that `POST /api/v1/media/episode/` always returns a 500 error:

```json
{"detail": "Internal Server Error.", "errors": "int() argument must be a string, a bytes-like object or a real number, not 'NoneType'"}
```

After reviewing `views.py`, the root cause is in `MediaTypeListView.post()`. The handler reads `season_number` from the request body but never reads `episode_number`:

```python
season_number = body.get("season_number")

try:
    metadata = services.get_media_metadata(
        media_type,
        media_id,
        source,
        [season_number],
    )
```

`episode_number` is never extracted from `body`, so it's always `None` when passed downstream to `get_media_metadata`, causing the `int()` cast to fail.

The fix would need to:
1. Extract `episode_number = body.get("episode_number")` from the body
2. Pass it to `get_media_metadata`
3. Include it in the `Item.objects.get_or_create` lookup fields

Tested with:
```bash
curl -X POST "https://{instance}/api/v1/media/episode/" \
  -H "Authorization: Bearer {token}" \
  -H "Content-Type: application/json" \
  -d '{"media_id": "387", "source": "tmdb", "season_number": 8, "episode_number": 18, "end_date": "2026-04-25"}'
```

`POST /api/v1/media/movie/` works correctly — this is specific to the `episode` (and likely `tv`) media types where child identifiers are needed.

This is the only issue I've encountered. otherwise everything related to movies and tv are working really well and I'm really impressed with the work you've done.

66Bunz commented:
That is because that route doesn't exists, but it should return a 404...
The season and episode endpoints are treated as children of a tv serie media, so to get to an episode you have to do `/api/v1/media/tv/[someid]/[season_number]/[episode_number]`

Haggler7039 commented:
> That is because that route doesn't exists, but it should return a 404... The season and episode endpoints are treated as children of a tv serie media, so to get to an episode you have to do `/api/v1/media/tv/[someid]/[season_number]/[episode_number]`

Thanks for the clarification. Just to make sure we're on the same page — in a previous comment you mentioned that `POST /api/v1/media/episode/` should be the correct way to track a new episode, and shared an example request body with `season_number` and `episode_number`. 

Is the intent still to support that endpoint, or has the design changed so that episode tracking should go through a different route? If the latter, what would the correct endpoint be for creating a new episode consumption record for an episode that doesn't already have a DB record?

For context, the use case is a scrobbler that needs to mark an episode as watched at playback completion. PATCH works fine for episodes that already exist in the DB (e.g. currently airing shows in the calendar), but fails with 404 for older episodes that have no existing record.

Here's what you said previously:
>Ah, nice, let me know how it'll go, if you're missing specific endpoints for the API, I haven't tested yet with a real implementation, so I don't know if something is missing yet.
>
> About tracking episodes, it theory the `POST /api/v1/media/{media_type}/`, but it's not working for episode, I have to check it because I think it was working before.
> 
> The end request should be something like:
> `POST /api/v1/media/episode/`
> Body:
> ```json
> {
>     "source": "tmdb",
>     "media_id": "61859",
>     "season_number": 2,
>     "episode_number": 3,
> }
> ```
> 
> The workflow for child elements (seasons/episodes) is not perfect yet, I was trying to make the parent medias work fine, sorry.
